### ISSUE TITLE
Multi-GPU training not starting or freezing. GPUs at 100%

### ISSUE THREAD

**Issue by Daniel451** [tensorflow/tensorflow#54325]
Very same script runs without issues on 1 GPU (setting NVIDIA_VISIBLE_DEVICES to one GPU) but freezes with 2 or 3 GPUs.

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu Server 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): 2.7 (Anaconda), 2.8 (official docker), 2.8 (self compiled docker)
- TensorFlow version (use command below): 2.8 & 2,7
- Python version: 3.9
- Bazel version (if compiling from source): 4.2.1
- GCC/Compiler version (if compiling from source): 9
- CUDA/cuDNN version: 11.4
- GPU model and memory: 3x RTX A6000 / 3x 48GB

**Describe the current behavior**
Training script hangs at executing `model.fit()`. Terminal shows `Loaded cuDNN version 8201` thrice (once for each GPU) and displays "TensorFloat-32 will be used..." once. GPUs turn to 100% but nothing gets calculated, CPU usage is nearly 0%, "Epoch 1/50" is displayed by `model.fit()` but no progress bar and nothing happens. Script is also unresponsive, even to Ctrl+C => process has to be killed manually.

**Describe the expected behavior**
After showing "Loaded cuDNN 8201" training should start and progress bar of `model.fit()` call should show up.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing): -

**Standalone code to reproduce the issue**
My code is a more complex scenario but even this simple code here shows the same issue (executes right away on 1 GPU but freezes with 2 or 3 GPUs):
```
import tensorflow as tf
import numpy as np
 
 
X = np.random.random((1000, 128, 128, 3)).astype(np.float32)
Y = np.random.random((1000, 10)).astype(np.float32)
 
dat_x = tf.data.Dataset.from_tensor_slices(X)
dat_y = tf.data.Dataset.from_tensor_slices(Y)
ds = tf.data.Dataset.zip((dat_x, dat_y))
ds = ds.batch(96)
ds = ds.repeat(50)
 
strategy = tf.distribute.MirroredStrategy()
 
with strategy.scope():
    inputs = tf.keras.Input(shape=(128, 128, 3))
    x = tf.keras.layers.Conv2D(32, (3, 3))(inputs)
    x = tf.keras.layers.MaxPool2D()(x)
    x = tf.keras.layers.Conv2D(32, (3, 3))(x)
    x = tf.keras.layers.MaxPool2D()(x)
    x = tf.keras.layers.Flatten()(x)
    x = tf.keras.layers.Dense(10)(x)
    model = tf.keras.Model(inputs=inputs, outputs=x)
    optim = tf.optimizers.Adam()
 
    model.compile(optim, loss=tf.keras.losses.CategoricalCrossentropy())
 
 
model.fit(ds, epochs=50)

```

**Comment by jiayugedede** [comment]
I have the same problems, I have checked the driver(ok), cuda(ok), cudnn(ok), nccl(ok), all of these were normal operation. However, the distribution training was broken and frozen. It is so disappointing.

**Comment by jurahul** [comment]
Do you have any logs that I can look into?

**Comment by lcorcodilos** [comment]
Commenting that I'm seeing the same behavior though with `cuDNN version 8400`

**Comment by Daniel451** [comment]
After lots of headache and different attempts I am suspecting that it actually has to do with the cuda architecture in use and the OS version / driver version / linux kernel version.

In my case, nothing helped to get multiple GPUs to work other than updating to `Ubuntu 22.04`, ensuring that a new linux kernel `5.15` is running _and_ having `CUDA >= 11.3` + updating to latest driver.

I guess for new(er) GPUs, as mentioned I was using 3x RTX A6000, thus Ampere generation NVIDIA GPUs, there is something that simply breaks with `Ubuntu 20.04` and/or the old linux kernel available in aforementioned Ubuntu. Even an `Ubuntu 22.04` in a docker environment was not helping.

So TL;DR:
For me updating to a more modern Ubuntu, updating linux kernel and _then_ using `CUDA >= 11.3` and latest nvidia driver worked.

Any other attempt, like different CUDA versions, driver versions, with or without docker, with Anaconda, ... did not work under `Ubuntu 20.04` and older linux kernel with new NVIDIA Ampere cards.

**Comment by jayagami** [comment]
This could be a nccl problem.

https://github.com/NVIDIA/nccl-tests

Try 

```
NCCL_DEBUG=INFO ./build/all_reduce_perf -b 8 -e 128M -f 2 -g 3
```

and you will get the same result

**Comment by aminfardi-CD** [comment]
Is there a solution to this? Running into the same issue and all my packages are upgraded as suggested.

**Comment by jurahul** [comment]
Assigning to @rohan100jain to find the right owner for this issue.

**Comment by WissamAntoun** [comment]
I had an issue similar to this, and i fixed it by disabling IOMMU in the bios. This is relevant if you have an AMD CPU and using Data Parallel

**Comment by aminfardi-CD** [comment]
I got around the issue for now by modifying the `cross_device_ops` to `tf.distribute.HierarchicalCopyAllReduce()` as per [stackoverflow ](https://stackoverflow.com/questions/59993176/tensorflow-mirroredstrategy-not-working-for-multi-gpu-training). Full code snippet:

`strategy = tf.distribute.MirroredStrategy( cross_device_ops=tf.distribute.HierarchicalCopyAllReduce())`

I still think this is an issue with the nccl, because the default argument for that parameter is `tf.distribute.NcclAllReduce`

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
